File: lib/py/uapi/internal/schema/CatchErrorCollisions.py
from typing import cast


def catch_error_collisions(u_api_schema_name_to_pseudo_json: dict[str, list[object]], error_keys: set[str], keys_to_index: dict[str, int], schema_keys_to_document_name: dict[str, str], document_names_to_json: dict[str, str]) -> None:
    from uapi.UApiSchemaParseError import UApiSchemaParseError
    from uapi.internal.schema.SchemaParseFailure import SchemaParseFailure
    from uapi.internal.schema.GetPathDocumentCoordinatesPseudoJson import get_path_document_coordinates_pseudo_json

    parse_failures = []
    error_keys_list = list(error_keys)
    error_keys_list.sort(key=lambda k: (
        schema_keys_to_document_name[k], keys_to_index[k]))

    for i in range(len(error_keys_list)):
        for j in range(i + 1, len(error_keys_list)):
            def_key = error_keys_list[i]
            other_def_key = error_keys_list[j]

            index = keys_to_index[def_key]
            other_index = keys_to_index[other_def_key]

            document_name = schema_keys_to_document_name[def_key]
            other_document_name = schema_keys_to_document_name[other_def_key]

            u_api_schema_pseudo_json = u_api_schema_name_to_pseudo_json[document_name]
            other_u_api_schema_pseudo_json = u_api_schema_name_to_pseudo_json[other_document_name]

            def_ = cast(dict[str, object], u_api_schema_pseudo_json[index])
            other_def = cast(dict[str, object],
                             other_u_api_schema_pseudo_json[other_index])

            err_def = cast(list[object], def_[def_key])
            other_err_def = cast(list[object], other_def[other_def_key])

            for k in range(len(err_def)):
                this_err_def = cast(dict[str, object], err_def[k])
                this_err_def_keys = set(this_err_def.keys())
                this_err_def_keys.discard("///")

                for l in range(len(other_err_def)):
                    this_other_err_def = cast(
                        dict[str, object], other_err_def[l])
                    this_other_err_def_keys = set(this_other_err_def.keys())
                    this_other_err_def_keys.discard("///")

                    if this_err_def_keys == this_other_err_def_keys:
                        this_error_def_key = next(iter(this_err_def_keys))
                        this_other_error_def_key = next(
                            iter(this_other_err_def_keys))
                        final_this_path = [
                            index, def_key, k, this_error_def_key]
                        final_this_document_json = document_names_to_json[document_name]
                        final_this_location_pseudo_json = get_path_document_coordinates_pseudo_json(
                            final_this_path, final_this_document_json)
                        parse_failures.append(SchemaParseFailure(
                            other_document_name,
                            [other_index, other_def_key, l,
                                this_other_error_def_key],
                            "PathCollision",
                            {"document": document_name,
                             "path": final_this_path,
                             "location": final_this_location_pseudo_json}
                        ))

    if parse_failures:
        raise UApiSchemaParseError(parse_failures, document_names_to_json)

File: lib/java/src/main/java/uapi/internal/schema/CatchErrorCollisions.java
package uapi.internal.schema;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import uapi.UApiSchemaParseError;

public class CatchErrorCollisions {
    public static void catchErrorCollisions(Map<String, List<Object>> uApiSchemaNameToPseudoJson, Set<String> errorKeys,
            Map<String, Integer> keysToIndex, Map<String, String> schemaKeysToDocumentName,
            Map<String, String> documentNamesToJson) {
        List<SchemaParseFailure> parseFailures = new ArrayList<>();

        var errorKeysList = new ArrayList<>(errorKeys);

        errorKeysList.sort((k1, k2) -> {
            var documentName1 = schemaKeysToDocumentName.get(k1);
            var documentName2 = schemaKeysToDocumentName.get(k2);
            if (!documentName1.equals(documentName2)) {
                return documentName1.compareTo(documentName2);
            } else {
                var index1 = keysToIndex.get(k1);
                var index2 = keysToIndex.get(k2);
                return index1 - index2;
            }
        });

        for (var i = 0; i < errorKeysList.size(); i++) {
            for (var j = i + 1; j < errorKeysList.size(); j++) {
                var defKey = errorKeysList.get(i);
                var otherDefKey = errorKeysList.get(j);

                int index = keysToIndex.get(defKey);
                int otherIndex = keysToIndex.get(otherDefKey);

                var documentName = schemaKeysToDocumentName.get(defKey);
                var otherDocumentName = schemaKeysToDocumentName.get(otherDefKey);

                var uApiSchemaPseudoJson = uApiSchemaNameToPseudoJson.get(documentName);
                var otherUApiSchemaPseudoJson = uApiSchemaNameToPseudoJson.get(otherDocumentName);

                var def = (Map<String, Object>) uApiSchemaPseudoJson.get(index);
                var otherDef = (Map<String, Object>) otherUApiSchemaPseudoJson.get(otherIndex);

                var errDef = (List<Object>) def.get(defKey);
                var otherErrDef = (List<Object>) otherDef.get(otherDefKey);

                for (int k = 0; k < errDef.size(); k++) {
                    var thisErrDef = (Map<String, Object>) errDef.get(k);
                    var thisErrDefKeys = new HashSet<>(thisErrDef.keySet());
                    thisErrDefKeys.remove("///");

                    for (int l = 0; l < otherErrDef.size(); l++) {
                        var thisOtherErrDef = (Map<String, Object>) otherErrDef.get(l);
                        var thisOtherErrDefKeys = new HashSet<>(thisOtherErrDef.keySet());
                        thisOtherErrDefKeys.remove("///");

                        if (thisErrDefKeys.equals(thisOtherErrDefKeys)) {
                            String thisErrorDefKey = thisErrDefKeys.iterator().next();
                            String thisOtherErrorDefKey = thisOtherErrDefKeys.iterator().next();
                            List<Object> finalThisPath = List.of(index, defKey, k, thisErrorDefKey);
                            var finalThisDocument = documentNamesToJson.get(documentName);
                            var finalThisLocation = GetPathDocumentCoordinatesPseudoJson
                                    .getPathDocumentCoordinatesPseudoJson(finalThisPath, finalThisDocument);
                            parseFailures.add(new SchemaParseFailure(
                                    otherDocumentName,
                                    List.of(otherIndex, otherDefKey, l, thisOtherErrorDefKey),
                                    "PathCollision",
                                    Map.of("document", documentName, "path",
                                            finalThisPath, "location", finalThisLocation)));
                        }
                    }
                }
            }
        }

        if (!parseFailures.isEmpty()) {
            throw new UApiSchemaParseError(parseFailures, documentNamesToJson);
        }
    }
}
